Dockerfile

```diff
@@ -6,12 +6,15 @@
 # docker build -t nav2:latest \
 #   --build-arg UNDERLAY_MIXINS \
 #   --build-arg OVERLAY_MIXINS ./
-ARG FROM_IMAGE=osrf/ros2:testing
 ARG UNDERLAY_WS=/opt/underlay_ws
 ARG OVERLAY_WS=/opt/overlay_ws
 
+# multi-stage for dependabot
+# https://github.com/dependabot/dependabot-core/issues/2057
+FROM osrf/ros2:testing-20210605003201 AS from_image
```

Comment:
Set to a specific date?

Reply:
Fixing the tag (to the date + utc time as proposed in https://github.com/osrf/docker_images/pull/546 ) will allow dependabot to notify us that the parent image has updated. When merging dependabot's PRs, we can determistcily break the build cache to ensure our CI image is built using the latest image layers from the upstream parent image registry repo.

Reply:
We'd have to update this every time there's a testing sync update? That's alot of overhead to maintain long-term.

Reply:
The `osrf/ros2:testing` image updates with the same cadence as the `library/ubuntu:20.04` image, so about once or twice every two months. So the overhead is essentially one of us clicking the merge button once or twice every two months. Doesn't seem like a lot to maintain for a rigorous and reproducible build environment.

For example, if someone pushes a breaking change or ABI update to one of our dependencies in the testing apt repo, we could just roll back and pin to the last working CI image tag until a fix is synced, unblocking our CI jobs in the meantime.

Or, let's say we're triaging an issue and we want to determine when the issue first surfaced by building against previous versions of our package dependencies. Because the apt testing repo only host's the most recent sync, we could still go back to previous CI image tags to access and build and test against these prior dependency versions.

For me personally though, when I'm debugging an issue with the CI, I'd like to be able to locally build and test against the exact same docker image that the CI used for a particular workflow run. One can use the image SHA digest for this, but timestamp tags are a little nicer to work with, as they are self chronological and still don't shift underneath you.

Reply:
What about people that want to fork nav2 and work on it? They wouldn't get any of these updates. What about updates to existing nav2 distribution branches like Foxy and Galactic? Are we also going to specifically tag images for those as well? 

>  So the overhead is essentially one of us clicking the merge button once or twice every two months. Doesn't seem like a lot to maintain for a rigorous and reproducible build environment.

I agree and I get the point that its nice to explicitly know when things changed for debugging. However this statement:

> For example, if someone pushes a breaking change to one of our dependencies in the testing apt repo, we could just roll back and pin to the last working CI image tag until a fix is synced, unblocking our CI jobs in the meantime.

is concerning to me, because testing changes are bound for the distribution release. Hopefully its patched, but we have to be aware of these bugs and somewhat roll with the punches because if its not patched then it will be given out to users in the next distribution sync! 

I agree its nice for us to know when things change, but honestly if we think there's going to be these many issues, I don't think we should be using `testing` at all. Maybe we should just use rolling. I know that was an issue due to TF/rclcpp issues in the aftermath of Galactic, but I don't see why we don't revert to that now so that folks can `apt install` the dependencies for Nav2 for development into `main`.

I'm concerned at this level of specificity for 1) developers 2) forkers 3) existing released distribution policies

Reply:
> What about people that want to fork nav2 and work on it? They wouldn't get any of these updates that are otherwise compiled intrinsically.

You may be confusing two things.
This PR does pin the parent image tag for our CI image, but it does not pin the tag for the CI image (though one could modify our CI's executors config to do that). I.e pinning the parent image tag vs pinning the CI image tag:

* https://github.com/ruffsl/navigation2/blob/757e1853f1c7ba1cfab62493c6d1f11b9591b86c/Dockerfile#L14

* https://github.com/ruffsl/navigation2/blob/757e1853f1c7ba1cfab62493c6d1f11b9591b86c/.github/workflows/update_ci_image.yaml#L102-L104

Forks who don't modify the CI's executors config would still pull from the same latest and greatest `navigation2:main` tag:

https://github.com/ruffsl/navigation2/blob/757e1853f1c7ba1cfab62493c6d1f11b9591b86c/.circleci/config.yml#L436-L439

> What about updates to existing nav2 distribution branches like Foxy and Galactic? Are we also going to specifically tag images for those as well?

No, we aren't currently building CI images for previous release branches. I had hope to, but CircleCI doesn't yet have a workflow config syntax to make that simple to implement. 

https://discuss.circleci.com/t/conditioning-docker-image-via-base-branch-name-from-pr/28426

Perhaps GitHub actions may facilitate this better after a few version iterations of their workflow config syntax. For production use, I always encourage users to build FROM the official `library/ros:latest` images and apt install nav2 like normal.

> Maybe we should just use rolling. ... Was there some reason we went away from the released rolling distribution other than the issues we were having with TF/rclcpp at the time? I don't recall.

Well, we moved the parent image for CI from `ros:rolling` to `osrf/ros2:testing` at the behest of the core ROS2 team from an offline email chain about the aftermath of Galactic. But this touches back to something you've just mentioned here:

> but we have to be aware of these bugs and somewhat roll with the punches because if its not patched then it will be given out to users in the next distribution sync!

Having our CI run against rolling testing sync allows us to catch breakages early on before the next distribution sync.

> It saves our CI at the expense of actual user issues in the field using the debians.

However, this won't be what the current PR configures. I only hypothesized changing the CI's executors config to pull from a pinned CI image tag as a last resort, incase someone wanted to continue running tests for a WIP feature PR unrelated to proper patch fix PR. It's not something I think would be commonly used, just thinking out loud for posterity.

Reply:
>  This PR does pin the parent image tag for our CI image, but it does not pin the tag for the CI image (though one could modify our CI's executors config to do that). I.e pinning the parent image tag vs pinning the CI image tag:

If its `FROM osrf/ros2:testing-20210605003201 AS from_image` and I fork it and never update the docker image, it will forever be `FROM osrf/ros2:testing-20210605003201 AS from_image` every time I want to build a docker image using it (unless I'm a vigilant enough user to notice it, which I wouldn't count on, on average). A forked user will never get the updated the base image unless they keep really extra special care to? 

I'm wasn't mentioning about forked users's CI experience versus main repo CI experiences,  But that does bring up the good question of what happens if a user on a forked branch submits a PR, does it use _our_ version that might be up to date or _their_ version that could be months old? I don't see any `ARG`s in the dockerfile that would let our CI override to the main branch's current default. 

---

> Well, we moved the parent image for CI from ros:rolling to osrf/ros2:testing at the behest of the core ROS2 team from an offline email chain about the aftermath of Galactic.

Admittedly I'm juggling a number of balls so my memory on any given email chain is a little fuzzy, but I thought that the main reason we asked about that was because they didn't want to do regular and predictable syncs to Rolling -- which was more of an extremely helpful suggestion than a critical need. 

The core of my concern is that we're making alot more work for ourselves. I can say pretty confidently I'm not going to have any more free time in the future than I've had historically to manage CI related issues so if anything here breaks, I'm probably not going to be able to fix it. So something relatively simple and straight forward is something I prefer. Tagging specific images seems like it might create a bunch of problems and I'm not exactly sure that the benefits of being technically correct outweigh the additional maintenance effort / complexity. I pride myself on making Nav very approachable to many people and the more esoteric we make the CI or local developer installs, it steers away from my intended mission of making it easy to contribute and use Nav2 for experts and newbies alike. 

The "nightly" builds I could understand better and I think make some intuitive sense to people. The rolling-testing is 1 somewhat hard to explain and new concept to many users (`rolling`) buried in another (`shadow-fixed` / `testing`) that primarily only core maintainers are even aware of, let alone deeply familiar with. 

So I'd just like us to take a step back and think about the trade offs before going too far down this rabbit hole. I agree with you 100% that this is workable and is probably the most technically correct option to make sure we track everything finely. If we were a company and I had a dev ops team dedicated to these kinds of things, I would agree this is the right course of action. But if we run CI using this system, we'll need to tell developers to do the same so that the environments are the same and they can find issues to resolve that are displayed in CI. As such, the complexity we can make our CI versioning system has to be balanced with what we can reasonably expect developers to put up with to help out. As far as I know, we have 3 major sets of options:

- Nightly: Pro latest and greatest, Con no debians for easy developer install / use, can be problematic if things break upstream of us that a release cycle would at least shield us from
- Rolling: Pro debians can be easily installed for developers to develop and debians if researchers want to use this on their robots as if it were a released distribution to use the latest and greatest of Nav2's features/bug fixes, Con not updated as often as we'd like so we might not see any of the API/ABI breaking changes right up until a new distribution release if other maintainers aren't regularly releasing to it. The months around distribution releases are problematic and stressful. 
- Rolling/testing: Pro complete control over dates released so we can roll back if something breaks and we get any new changes to dependencies 2-3 weeks before in Rolling proper to resolve any issues, get debians as well, Con have to explain 2 somewhat complex concepts to new users to be able to understand what they're doing just to build the codebase. Can't install rolling next to Rolling-testing.

Reply:
> A forked user will never get the updated the base image unless they keep really extra special care to?

Fair enough, I've fixed the Dockerfile to prevent this. Developers may continue using the default unpinned development stage, while our CI may continue using the pinned integration. See 0c221cd4990905ad33570236c5862789700b6e86

A reason I'd like to keep with the dependabot approach is that there is no other easy way of determining if a parent image layer is outdated in the CI image unless we rebuild the image. I'd like to insure the CI image never gets too stale without rebuilding the entire thing every day. We've looked for alternatives, but they aren't nearly as elegant or minimal as just enabling dependabot:

https://github.com/osrf/docker_images/issues/362

> But that does bring up the good question of what happens if a user on a forked branch submits a PR, does it use our version that might be up to date or their version that could be months old?

The docker image tags selected in the CI executor config are fully qualified, and will thus pull our version of the CI images to test the PR. Even if they opened a PR internal to their own repo, it would still use our version of the CI images unless they manually modified the CI config to something else:

``` yaml
executors:
  debug_exec:
    docker:
      - image: ghcr.io/ros-planning/navigation2:main
# https://github.com/ruffsl/navigation2/blob/757e1853f1c7ba1cfab62493c6d1f11b9591b86c/.circleci/config.yml#L439
```

---

> So I'd just like us to take a step back and think about the trade offs before going too far down this rabbit hole. I agree with you 100% that this is workable and is probably the most technically correct option to make sure we track everything finely.

I'll get back to you on this with a longer write up with my thoughts tomorrow, but I agree keeping with `ros:rolling` would have been my ideal choice. I'm looking into ways we can ensure even the official ros library images can be better kept up-to-date over at https://github.com/osrf/docker_images , such as ensuring syncs trigger rebuilds for the library images, and a version tagged so we can leverage dependant without needing to use the `osrf/ros2:testing` images.